src/services/sqlTasksService.ts

```diff
@@ -0,0 +1,179 @@
+     */
+    private handleTaskCreatedNotification(taskInfo: TaskInfo): void {
+        // Default to no-op for the progressCallback since we don't have the progress callback from the notification yet. There's
+        // potential here for a race condition in which the first update comes in before this callback is updated - if that starts
+        // happening then we'd want to look into keeping track of the latest update message to display as soon as the progress
+        // callback is set such that we update the notification correctly.
+        const newTaskInfo: ActiveTaskInfo = {
+            taskInfo,
+            progressCallback: () => { return; },
+            completionPromise: new Deferred<void>()
+        };
+
+        vscode.window.withProgress(
+            {
+                location: vscode.ProgressLocation.Notification,
+                title: taskInfo.name,
+                cancellable: false
+            },
+            async (progress, _token): Promise<void> => {
+                newTaskInfo.progressCallback = value => progress.report(value);
+                await newTaskInfo.completionPromise;
+            }
+        );
+        this._activeTasks.set(taskInfo.taskId, newTaskInfo);
+    }
+
+    /**
+     * Handles an update to an existing task, updating the current progress notification as needed with any new
+     * status/messages. If the task is completed then completes the progress notification and displays a final toast
+     * informing the user that the task was completed.
+     * @param taskProgressInfo The progress info for the task
+     */
+    private handleTaskChangedNotification(taskProgressInfo: TaskProgressInfo): void {
+        const taskInfo = this._activeTasks.get(taskProgressInfo.taskId);
+        if (!taskInfo) {
+            console.warn(`Status update for unknown task ${taskProgressInfo.taskId}`!);
+            return;
+        }
+        const taskStatusString = toTaskStatusDisplayString(taskProgressInfo.status);
+        if (isTaskCompleted(taskProgressInfo.status)) {
+            // Task is completed, complete the progress notification and display a final toast informing the
+            // user of the final status.
+            this._activeTasks.delete(taskProgressInfo.taskId);
+            taskInfo.completionPromise.resolve();
+            // Only include the message if it isn't the same as the task status string we already have - some (but not all) task status
+            // notifications include this string as the message
+            const taskMessage = taskProgressInfo.message && taskProgressInfo.message.toLowerCase() !== taskStatusString.toLowerCase() ?
+                utils.formatString(localizedConstants.taskStatusWithMessage, taskInfo.taskInfo.name, taskStatusString, taskProgressInfo.message) :
+                utils.formatString(localizedConstants.taskStatusWithName, taskInfo.taskInfo.name, taskStatusString);
+            vscode.window.showInformationMessage(taskMessage);
```

Comment:
it would be nice if there was also showErrorMessage() if the TaskStatus was failed

Reply:
Good idea!